### PR TITLE
Fix bump mapping

### PR DESCRIPTION
**Pull request by dictoon**
- Fix built-in bump mapping algorithm
- Allow using bump mapping and normal mapping with non-texture sources
- Review, check and cleanup code computing screen space derivatives dpdx, dpdy, duvdx and duvdy
- Reimplement appleseed-max's bump mapping OSL shader
- Remove parameter attributes from appleseed-max's normal mapping OSL shader
- Add test scenes

### CODE REVIEW

File: src/appleseed/renderer/modeling/material/material.cpp [hunk 1]
```diff
@@ -228,32 +228,25 @@ IBasisModifier* Material::create_basis_modifier(const MessageContext& context) c
...
-    // Retrieve the displacement texture.
-    const TextureSource* displacement_map = static_cast<const TextureSource*>(displacement_source);
-    const Texture& texture = displacement_map->get_texture_instance().get_texture();
+    if (displacement_source == nullptr)
+        return nullptr;
 
-    // Print a warning if the displacement texture is not expressed in the linear RGB color space.
-    if (texture.get_color_space() != ColorSpaceLinearRGB)
+    // Additional checks if a texture instance is used for displacement.
+    const TextureSource* texture_map = dynamic_cast<const TextureSource*>(displacement_source);
```
Comment (usakhelo) [inline src/appleseed/renderer/modeling/material/material.cpp]: Franz, I think it is still looking for TextureSource here?

Comment (dictoon) [reply src/appleseed/renderer/modeling/material/material.cpp]: Yes that's correct, it's checking if the source is a texture source and if so, does an additional check.

Comment (usakhelo) [reply src/appleseed/renderer/modeling/material/material.cpp]: Nevermind, it's something else. I'm still checking.

Comment (usakhelo) [reply src/appleseed/renderer/modeling/material/material.cpp]: Something is wrong, still checking..:
image

Comment (dictoon) [reply src/appleseed/renderer/modeling/material/material.cpp]: I see. In the appleseed code, try decommenting the line that says

    //#define USE_SCREEN_SPACE_UV_DERIVATIVES

on line 53 of `src\appleseed\renderer\modeling\material\bumpmappingmodifier.cpp`.

Comment (dictoon) [reply src/appleseed/renderer/modeling/material/material.cpp]: That's an alternative bump mapping method that may generally work better.

**Comment by usakhelo** [comment]
Franz, I've applied bump material similar to what was already in scene test scene to a sphere and it breaks the smoothing. Attached packed project renamed to .zip.

image


[test_smooth_bump.zip](https://github.com/appleseedhq/appleseed/files/1312695/test_smooth_bump.zip)

**Comment by dictoon** [comment]
Thanks @usakhelo, I'm looking into it.

**Pull request merged**